### PR TITLE
PBD-102 Fix escrow commands [reupload]

### PR DESCRIPTION
**Pull request by CryptoCubik**
Refactor commands, fix strings, update tests, write negative tests.

### CODE REVIEW

File: src/api/commands/escrow/EscrowReleaseCommand.ts [hunk 1]
```diff
@@ -62,27 +68,35 @@ export class EscrowReleaseCommand extends BaseCommand implements RpcCommandInter
     public async validate(data: RpcRequest): Promise<RpcRequest> {
 
         if (data.params.length < 2) {
-            throw new MessageException('Missing params.');
+            throw new MessageException('Missing params orderItemId and memo.');
```
Comment (xludx) [inline src/api/commands/escrow/EscrowReleaseCommand.ts]: use MissingParamException

File: src/api/commands/escrow/EscrowReleaseCommand.ts [hunk 2]
```diff
@@ -62,27 +68,35 @@ export class EscrowReleaseCommand extends BaseCommand implements RpcCommandInter
...
         }
 
-        const orderItemModel = await this.orderItemService.findOne(data.params[0]);
-        const orderItem = orderItemModel.toJSON();
+        const orderItemId = data.params[0];
+        if (typeof orderItemId !== 'number' || orderItemId < 0) {
+            throw new MessageException('orderItemId needs to be a number >= 0.');
```
Comment (xludx) [inline src/api/commands/escrow/EscrowReleaseCommand.ts]: use InvalidParamException

File: src/api/commands/escrow/EscrowRemoveCommand.ts [hunk 3]
```diff
@@ -36,21 +39,40 @@ export class EscrowRemoveCommand extends BaseCommand implements RpcCommandInterf
...
+            throw new MissingParamException('listingItemTemplateId');
         }
 
         const listingItemTemplateId = data.params[0];
+        if (typeof listingItemTemplateId !== 'number' || listingItemTemplateId < 0) {
+            throw new InvalidParamException('listingItemTemplateId', 'number');
+        }
+
         const listingItemTemplateModel = await this.listingItemTemplateService.findOne(listingItemTemplateId);
+        if (!listingItemTemplateModel) {
```
Comment (xludx) [inline src/api/commands/escrow/EscrowRemoveCommand.ts]: findOne throws NotFoundException

File: src/api/commands/escrow/EscrowLockCommand.ts [hunk 4]
```diff
@@ -95,13 +97,40 @@ export class EscrowLockCommand extends BaseCommand implements RpcCommandInterfac
...
 
+    public async validate(data: RpcRequest): Promise<RpcRequest> {
+        if (data.params.length >= 1) {
+            const orderItemId = data.params[0];
+            if (typeof orderItemId !== 'number') {
+                throw new MissingParamException('orderItemId must be number.');
+            }
+            const orderItemModel = await this.orderItemService.findOne(orderItemId);
+            if (!orderItemModel) {
+                throw new MessageException(`orderItemModel with orderItemId = <${orderItemId}> not found.`);
```
Comment (xludx) [inline src/api/commands/escrow/EscrowLockCommand.ts]: findOne will throw and this code is never reached.

```
await this.orderItemService.findOne(orderItemId)
    .catch(reason => {
        this.log.error('error: ', reason);
        throw new ModelNotFoundException('OrderItem');
    });

```

File: src/api/commands/escrow/EscrowRefundCommand.ts [hunk 5]
```diff
@@ -91,14 +59,79 @@ export class EscrowRefundCommand extends BaseCommand implements RpcCommandInterf
...
 
+    public async validate(data: RpcRequest): Promise<RpcRequest> {
+        let orderItem;
+        if (data.params.length >= 1) {
+            const orderItemId = data.params[0];
+            if (typeof orderItemId !== 'number' || orderItemId < 0) {
+                throw new InvalidParamException('orderItemId must be number and >= 0.', 'number');
+            }
+            const orderItemModel = await this.orderItemService.findOne(orderItemId);
+            if (!orderItemModel) {
```
Comment (xludx) [inline src/api/commands/escrow/EscrowRefundCommand.ts]: here again, findOne will allready throw NotFoundException

File: src/api/commands/escrow/EscrowRefundCommand.ts [hunk 6]
```diff
@@ -91,14 +59,79 @@ export class EscrowRefundCommand extends BaseCommand implements RpcCommandInterf
...
+            const escrow = orderItem.Bid.ListingItem.PaymentInformation.Escrow;
+            if (_.isEmpty(escrow)) {
+                this.log.error('Escrow not found!');
+                throw new MessageException('Escrow not found!');
+            }
+
+            const escrowRatio = orderItem.Bid.ListingItem.PaymentInformation.Escrow.Ratio;
+            if (_.isEmpty(escrowRatio)) {
+                this.log.error('EscrowRatio not found!');
+                throw new InvalidParamException('EscrowRatio not found!');
```
Comment (xludx) [inline src/api/commands/escrow/EscrowRefundCommand.ts]: ModelNotFoundException('EscrowRatio');

File: src/api/commands/escrow/EscrowRefundCommand.ts [hunk 7]
```diff
@@ -91,14 +59,79 @@ export class EscrowRefundCommand extends BaseCommand implements RpcCommandInterf
...
+            }
+        }
+        if (data.params.length >= 2) {
+            // TODO: Accepted status seems like something that needs validation as it sounds like an enum
+            const accepted = data.params[1];
+            if (typeof accepted !== 'boolean') {
+                throw new InvalidParamException('accepted must be boolean.', 'boolean');
```
Comment (xludx) [inline src/api/commands/escrow/EscrowRefundCommand.ts]: InvalidParamException('accepted', 'boolean');

File: src/api/commands/escrow/EscrowRefundCommand.ts [hunk 8]
```diff
@@ -91,14 +59,79 @@ export class EscrowRefundCommand extends BaseCommand implements RpcCommandInterf
...
+            }
+        }
+        if (data.params.length >= 3) {
+            const memo = data.params[2];
+            if (typeof memo !== 'string') {
+                throw new InvalidParamException('memo must be string.', 'string');
```
Comment (xludx) [inline src/api/commands/escrow/EscrowRefundCommand.ts]: InvalidParamException('memo', 'string');

File: src/api/commands/escrow/EscrowReleaseCommand.ts [hunk 9]
```diff
@@ -61,28 +67,38 @@ export class EscrowReleaseCommand extends BaseCommand implements RpcCommandInter
...
-            OrderStatus.ESCROW_LOCKED,
-            OrderStatus.SHIPPING
-        ];
+        const memo = data.params[1];
+        if (typeof memo !== 'string') {
+            throw new InvalidParamException('memo', 'string');
+        }
+
+        const orderItemModel = await this.orderItemService.findOne(orderItemId);
+        if (!orderItemModel) {
```
Comment (xludx) [inline src/api/commands/escrow/EscrowReleaseCommand.ts]: unnecessary

File: src/api/commands/escrow/EscrowRemoveCommand.ts [hunk 10]
```diff
@@ -36,21 +39,41 @@ export class EscrowRemoveCommand extends BaseCommand implements RpcCommandInterf
...
         }
 
         const listingItemTemplateId = data.params[0];
+        if (typeof listingItemTemplateId !== 'number' || listingItemTemplateId < 0) {
+            throw new InvalidParamException('listingItemTemplateId', 'number');
+        }
+
+        // Throws NotFoundException if not found
         const listingItemTemplateModel = await this.listingItemTemplateService.findOne(listingItemTemplateId);
+        /* if (!listingItemTemplateModel) {
```
Comment (xludx) [inline src/api/commands/escrow/EscrowRemoveCommand.ts]: remove, no need to leave commented out code here

File: test/black-box/escrow/EscrowAddCommand.test.ts [hunk 11]
```diff
@@ -82,9 +259,20 @@ describe('EscrowAddCommand', () => {
         expect(result.type).toBe(testData.type);
         expect(result.Ratio.buyer).toBe(testData.ratio.buyer);
         expect(result.Ratio.seller).toBe(testData.ratio.seller);
+
+        // Double check the object was created with the new values.
```
Comment (xludx) [inline test/black-box/escrow/EscrowAddCommand.test.ts]: all service.create()'s will call service.findOne() in the end returning the created model, so double checking is propably not necessary.

Comment (CryptoCubik) [reply test/black-box/escrow/EscrowAddCommand.test.ts]: I understand that, but I think that within tests more checks is better than less.
Expect the impossible, etc

File: src/api/commands/escrow/EscrowLockCommand.ts [hunk 12]
```diff
@@ -95,13 +97,44 @@ export class EscrowLockCommand extends BaseCommand implements RpcCommandInterfac
 
     }
 
+    public async validate(data: RpcRequest): Promise<RpcRequest> {
+        if (data.params.length >= 1) {
+            const orderItemId = data.params[0];
+            if (typeof orderItemId !== 'number') {
+                throw new MissingParamException('orderItemId must be number.');
```
Comment (xludx) [inline src/api/commands/escrow/EscrowLockCommand.ts]: InvalidParamException

File: src/api/commands/escrow/EscrowLockCommand.ts [hunk 13]
```diff
@@ -95,13 +97,44 @@ export class EscrowLockCommand extends BaseCommand implements RpcCommandInterfac
...
+            }
+
+            let orderItemModel;
+            try {
+                orderItemModel = await this.orderItemService.findOne(orderItemId);
+            } catch (ex) {
```
Comment (xludx) [inline src/api/commands/escrow/EscrowLockCommand.ts]: no need for try catch

File: src/api/commands/escrow/EscrowLockCommand.ts [hunk 14]
```diff
@@ -95,13 +97,44 @@ export class EscrowLockCommand extends BaseCommand implements RpcCommandInterfac
...
+                this.log.error('Error: ' + ex);
+                throw new NotFoundException(orderItem);
+            }
+        }
+        if (data.params.length >= 2) {
+            const nonce = data.params[1];
+            if (typeof nonce !== 'string') {
+                throw new InvalidParamException('nonce must be string.', 'string');
```
Comment (xludx) [inline src/api/commands/escrow/EscrowLockCommand.ts]: throw new InvalidParamException('nonce', 'string');

File: src/api/commands/escrow/EscrowLockCommand.ts [hunk 15]
```diff
@@ -95,13 +97,44 @@ export class EscrowLockCommand extends BaseCommand implements RpcCommandInterfac
...
+            }
+        }
+        if (data.params.length >= 3) {
+            const memo = data.params[2];
+            if (typeof memo !== 'string') {
+                throw new InvalidParamException('memo must be string.', 'string');
```
Comment (xludx) [inline src/api/commands/escrow/EscrowLockCommand.ts]: throw new InvalidParamException('memo', 'string');

File: src/api/commands/escrow/EscrowRefundCommand.ts [hunk 16]
```diff
@@ -91,14 +59,83 @@ export class EscrowRefundCommand extends BaseCommand implements RpcCommandInterf
...
+        let orderItem;
+        if (data.params.length >= 1) {
+            const orderItemId = data.params[0];
+            if (typeof orderItemId !== 'number' || orderItemId < 0) {
+                throw new InvalidParamException('orderItemId', 'number');
+            }
+
+            let orderItemModel;
+            try {
+                orderItemModel = await this.orderItemService.findOne(orderItemId);
```
Comment (xludx) [inline src/api/commands/escrow/EscrowRefundCommand.ts]: no need for try catch

File: src/api/commands/escrow/EscrowReleaseCommand.ts [hunk 17]
```diff
@@ -61,28 +67,41 @@ export class EscrowReleaseCommand extends BaseCommand implements RpcCommandInter
...
-            OrderStatus.ESCROW_LOCKED,
-            OrderStatus.SHIPPING
-        ];
+        const memo = data.params[1];
+        if (typeof memo !== 'string') {
+            throw new InvalidParamException('memo', 'string');
+        }
+
+        let orderItemModel;
+        try {
```
Comment (xludx) [inline src/api/commands/escrow/EscrowReleaseCommand.ts]: no need for try catch

File: src/api/commands/escrow/EscrowRefundCommand.ts [hunk 18]
```diff
@@ -91,14 +59,83 @@ export class EscrowRefundCommand extends BaseCommand implements RpcCommandInterf
...
+                this.log.error('Escrow not found!');
+                throw new ModelNotFoundException('escrow');
+            }
+
+            const escrowRatio = orderItem.Bid.ListingItem.PaymentInformation.Escrow.Ratio;
+            if (_.isEmpty(escrowRatio)) {
+                this.log.error('EscrowRatio not found!');
+                throw new ModelNotFoundException('escrowRatio');
+            }
+        }
```
Comment (xludx) [inline src/api/commands/escrow/EscrowRefundCommand.ts]: missing else, should throw missingparamexception

Comment (xludx) [reply src/api/commands/escrow/EscrowRefundCommand.ts]: also, there's no need for the if else, data.params[0] should always exist and all the checks inside the if should always be run, the if just adds complexity which just caused a bug.

File: src/api/commands/escrow/EscrowRemoveCommand.ts [hunk 19]
```diff
@@ -36,21 +39,44 @@ export class EscrowRemoveCommand extends BaseCommand implements RpcCommandInterf
...
 
         const listingItemTemplateId = data.params[0];
-        const listingItemTemplateModel = await this.listingItemTemplateService.findOne(listingItemTemplateId);
+        if (typeof listingItemTemplateId !== 'number' || listingItemTemplateId < 0) {
+            throw new InvalidParamException('listingItemTemplateId', 'number');
+        }
+
+        // Throws NotFoundException if not found
+        let listingItemTemplateModel;
+        try {
```
Comment (xludx) [inline src/api/commands/escrow/EscrowRemoveCommand.ts]: no need for try catch

**Pull request closed without merging**